rfd/0050-join-methods.md

```diff
@@ -0,0 +1,117 @@
+
+## Why
+
+This RFD can serve as a reference for all of the ways Teleport nodes can get
+cluster credentials in order to facilitate discussion of the current methods and
+new endpoints that will be added for the IAM join method and Certificate Bot.
+
+## Details
+
+When "joining" a cluster, a Teleport node must authenticate itself using either
+a secret token or one of the new AWS join methods (EC2 or IAM method). If the
+node can successfully authenticate, the Auth server will return signed SSH and
+TLS certificates to the node for the requested role (Proxy, Node, Kubernetes,
+etc). The token or AWS join method is only used once for the first-time connect
+and from that point on signed certificates are used.
+
+The joining node also needs to be able to trust that the Teleport cluster is
+authentic. It can do this by using CA pins configured on the node, in the case
+of IoT nodes connecting to the proxy, the proxy TLS cert can be trusted through
+PKI.
+
+### Current Join Endpoints
+
+#### Auth `POST /tokens/register`
+Join methods: token, EC2
+Server trust: CA pins
+This is the main token register endpoint on the Auth server. Accepts tokens or
+EC2 Identity Documents to authenticate the client.
+
+#### Auth `POST /tokens/register/auth`
+Join Methods: token
+Server trust: CA pins
+Looks like this is a legacy endpoint the just deletes the given token. Only
+works for tokens with the `Auth` role. I can't find anything that calls this, so
+it can probably be scheduled for deletion.
+
+#### Auth `rpc GenerateHostCerts(HostCertsRequest) returns (Certs)`
+Client trust: mTLS
+Server trust: mTLS
+This is an mTLS authenticated gRPC endpoint for cert renewal. The node must
+first join the cluster using another endpoint to get its first client
+certificate used for the mTLS connection.
+
+#### Proxy `POST /webapi/host/credentials`
+Join methods: token, EC2
+Server trust: PKI
+This is the proxy endpoint for registering IoT nodes that don't have a direct connection to the auth server, it basically forwards to
+the Auth `/tokens/register` endpoint.
+
+### IAM Join Method
+The IAM join method (introduced in RFD 41) requires gRPC methods rather than
+HTTP because the design requires streams to implement a challenge/response
+protocol. To summarize, after the client initiates the rpc the Auth server sends
+a crypto-random "challenge" string over the server->client stream. This
+challenge must be included in a signed `sts:GetCallerIdentity` request which
+will be sent on the client->server stream.  Finally, if everything checks out,
+the Auth server will send the signed certificates on the server->client stream.
+
+A new gRPC streaming method will be added on Auth and Proxy to support this. The
+client will be able to call the gRPC method with either a Proxy or Auth address
+and it will "just work" transparently.
+
+#### Auth `rpc RegisterUsingIAM(stream RegisterUsingIAMRequest) returns (stream RegisterUsingIAMResponse)`
```

Comment:
Could we push all our joining/renewal needs to this new RPC family, regardless of the cert type? That could allow us to do a reasonably big API cleanup and tackle the various form of entity registering.

Reply:
That could be a good second phase to this, once we can do unauthenticated gRPC through the Proxy there's nothing stopping us. I will give the new gRPC service a pretty generic name to accommodate more authentication methods

Reply:
Don't be _too_ generic, as we don't want it to become a bag of anything. It's cheap to make new gRPC services and we can bind as many as we like to a single port.